aiida/backends/tests/tools/graph/test_age.py

```diff
@@ -0,0 +1,771 @@
+        queryb0 = QueryBuilder()
+        queryb0.append(Node, tag='t1')
+        queryb0.append(Node, with_incoming='t1')
+        uprule0 = UpdateRule(queryb0, max_iterations=2, track_edges=True)
+        uprule0_result = uprule0.run(basket0.copy())
+
+        result0 = uprule0_result['nodes'].keyset
+        expect0 = set(anode.id for _, anode in nodes.items())
+        self.assertEqual(result0, expect0)
+
+        result0 = set()
+        for data in uprule0_result['nodes_nodes'].keyset:
+            result0.add((data[0], data[1]))
+
+        expect0 = set()
+        expect0.add((nodes['DNI'].id, nodes['CN0'].id))
+        expect0.add((nodes['DNI'].id, nodes['WN1'].id))
+        expect0.add((nodes['DNI'].id, nodes['WN2'].id))
+        expect0.add((nodes['CN0'].id, nodes['DNO'].id))
+        expect0.add((nodes['WN1'].id, nodes['DNO'].id))
+        expect0.add((nodes['WN2'].id, nodes['DNO'].id))
+        expect0.add((nodes['WN2'].id, nodes['WN1'].id))
+        expect0.add((nodes['WN1'].id, nodes['CN0'].id))
+        self.assertEqual(result0, expect0)
+
+        # Backwards traversal (check partial traversal and link direction)
+        initial_nodes = [nodes['DNO'].id]
+        basket0 = Basket(nodes=initial_nodes)
+        queryb0 = QueryBuilder()
+        queryb0.append(Node, tag='t1')
+        queryb0.append(Node, with_outgoing='t1')
+        uprule0 = UpdateRule(queryb0, max_iterations=1, track_edges=True)
+        uprule0_result = uprule0.run(basket0.copy())
+
+        result0 = uprule0_result['nodes'].keyset
+        expect0 = set(anode.id for _, anode in nodes.items())
+        expect0 = expect0.difference(set([nodes['DNI'].id]))
+        self.assertEqual(result0, expect0)
+
+        result0 = set()
+        for data in uprule0_result['nodes_nodes'].keyset:
+            result0.add((data[0], data[1]))
+
+        expect0 = set()
+        expect0.add((nodes['CN0'].id, nodes['DNO'].id))
+        expect0.add((nodes['WN1'].id, nodes['DNO'].id))
+        expect0.add((nodes['WN2'].id, nodes['DNO'].id))
+        self.assertEqual(result0, expect0)
+
+    def test_empty_input(self):
+        """
+        Testing empty input.
+        """
+        input_nodes = []
+        basket0 = Basket(nodes=input_nodes)
+        queryb0 = QueryBuilder().append(Node).append(Node)
+        uprule0 = UpdateRule(queryb0, max_iterations=np.inf)
+        result0 = uprule0.run(basket0.copy())['nodes'].keyset
+        self.assertEqual(result0, set())
+
+
+class TestAiidaEntitySet(AiidaTestCase):
+    """Tests for AiidaEntitySets"""
+
+    def setUp(self):
+        super().setUp()
+        self.reset_database()
+
+    def test_class_mismatch(self):
+        """
+        Test the case that an AiidaEntitySet is trying to be used in an operation with another AiidaEntitySet
+            which containts mismatched aiida entities or in operation with a data type other than AiidaEntitySet
+        """
+        # pylint: disable=pointless-statement
```

Comment:
So, the problem is with these blocks:

```
with self.assertRaises(TypeError):
    aes_node + aes_group
```

Here we are testing that the correct error raises when we try to do algebra with mismatching types. My guess is that the pointless-statement raises because we are not assigning the result of the addition, but for what we are testing we don't really need to. What is the expected protocol here? Just assign it to a scratch (`_`) variable anyways?

Reply:
Not sure - I guess you could assign it to `_`, yeah.